pkg/runtime/process.go

```diff
@@ -0,0 +1,70 @@
+package runtime
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// Process controls the process execution
+type Process interface {
+	// Name returns the name of the process
+	Name() string
+	// Lock tries to acquire an execution lock to prevent concurrent executions.
+	// Returns error if lock is already acquired by another process.
+	Lock() error
+	// Unlock releases the execution lock
+	Unlock() error
+	// Profiler returns the process profiler
+	Profiler(ProfilerConfig) (Profiler, error)
```

Comment:
I'm not 100% sure on the interface here. I get `Lock` and `Unlock`, but the `Profiler` method seems to be just a proxy for the profiler constructor. Also not sure about what `Name` is used for.

If we make another implementation of this, which for the sake of the example could be a kubernetes-driven lock, why would that implementation need to also proxy the profiler constructor and return a name?

Reply:
> the Profiler method seems to be just a proxy for the profiler constructor. 

The main reason the profiler is a constructor method is that the configuration would come from CLI arguments passed to the command, we cannot just start/stop it.

> If we make another implementation of this ... why would that implementation need to also proxy the profiler constructor and return a name?

This is a good point. Another option would be to move it to the Runtime interface but I'm not sure that solves the issue.

> Also not sure about what Name is used for.
As for returning a Name, it is the process name, it prevents having to have access to `os.Args[0]` to get the name.

Reply:
My main point was not "these things should not exist", but rather "do these things actually belong to the same interface"? My point is that I don't see how the process name and the profiler are related to each other, so my first thought is whether we should split this.

Reply:
As mentioned before, at the end we need to have a single root for all interfaces, the runtime environment, which gives access to all other interfaces. So if we split it, we need to put it in another place. The question remains, where? 

As for the actual place, these methods (lock, profile) relates to controlling the process execution. As I expect it to have more functions in the future. For example, for implementing #117 I was expecting an interface or method for handling signals.

I agree that it may be debatable if the name belongs here, but again if not here, where?

Reply:
The comment was motivated by the principle of keeping interfaces small (I'm refraining from quoting Rob Pike here 😈). 

Here we are not using the interface to encapsulate behavior, we're doing so because the Go language enforces us to use interfaces if we want to mock things for testing.

> I agree that it may be debatable if the name belongs here, but again if not here, where?

I think whenever we are injecting this interface, we can inject others. Something along the lines of:

```go
rootCmd := commands.BuildRootCmd(Context{
    FSLock: locker.New(), // With default path
    ProcessInfo: processinfo.New(),
    Profiler: profiler.New(),
    Environment: env.New(),
})
```

That being said, that's my style of doing things, which it's not easy (or perhaps possible) to prove to be better than the current approach. We can keep this one and refactor it if it becomes a problem, or we reach an agreement on a new style.

Reply:
I Don't disagree with any of your points. And by the way, keep interfaces small predates go for at least one decade. 

I'm not sure what you are suggesting in the code above, `FSLock`, `ProcessInfo`, `Profiler`, and `Environment` are all interfaces, each with one method? 

Having many one-method interfaces that are commonly used together doesn't seem right to me either. Also, notice that in the current proposal the `Lock` method in the `Process` interface has a very concrete meaning: it acquires a lock for the process, it is not a generic lock mechanism (it uses a generic mechanism).

Reply:
> If I understand correctly, you are suggesting in the code above, FSLock, ProcessInfo, Profiler, and Environment are all interfaces, each with one method?

That's what I have in mind, yes. Some people prefer to make those function pointers, which is not very different. I think the best known example to this is the classic `func() time.Time`, which defaults to `time.Now`.

> Having many one-method interfaces that are commonly used together doesn't seem right to me either.

I agree with this, I don't think either approach is obviously better than another. When in doubt, I lean toward what the principles say, but I think your approach to lean towards the simpler is also very valid.

With no clear benefit from one versus the other this becomes another style discussion so I'm fine moving on with the current approach just for the sake of not blocking this further.